### PR TITLE
Remove Unpin constraints

### LINKED ISSUE #3136: Remove `Unpin` constraints where possible
**Issue summary**

Remove `Unpin` constraints for the `AsyncWriterWithChecksum` or other types if possible.

**Acceptance Criteria**

**Other information and links**

This was previously done [here](https://github.com/ChainSafe/forest/pull/3104/commits/5da089c13540609a18e7f4329f6eaf31ff0444dd) but we want to keep changes as small and “atomic” as possible.

### PR DESCRIPTION
**Pull request by elmattic**
## Summary of changes

Changes introduced in this pull request:

-

## Reference issue to close (if applicable)

Closes #3136

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/libp2p/rpc/decoder.rs [hunk 1]
```diff
@@ -44,31 +44,32 @@ where
...
+            let n = std::task::ready!(this.io.poll_read(cx, &mut buf))?;
             // Terminated
             if n == 0 {
                 let item = serde_ipld_dagcbor::de::from_reader(&self.bytes[..])
                     .map_err(|e| io::Error::new(io::ErrorKind::Other, e.to_string()));
                 return Poll::Ready(item);
             }
-            self.bytes_read += n;
-            if self.max_bytes_allowed > 0 && self.bytes_read > self.max_bytes_allowed {
+            *this.bytes_read += n;
```
Comment (LesnyRumcajs) [inline src/libp2p/rpc/decoder.rs:55]: Why do you use C++ in Rust?

Comment (elmattic) [reply src/libp2p/rpc/decoder.rs:55]: Haha, if only I could do `*this->bytes_read++;`

File: src/utils/io/writer_checksum.rs [hunk 2]
```diff
@@ -45,24 +46,24 @@ impl<D: Digest, W: AsyncWriteExt + Unpin> AsyncWrite for AsyncWriterWithChecksum
...
     }
 
     fn poll_close(
-        mut self: Pin<&mut Self>,
+        self: Pin<&mut Self>,
         cx: &mut std::task::Context<'_>,
     ) -> Poll<std::io::Result<()>> {
-        Pin::new(&mut self.inner).poll_close(cx)
+        let this = self.project();
+        this.inner.poll_close(cx)
```
Comment (lemmih) [inline src/utils/io/writer_checksum.rs:61]: nit: I'm partial to one-liners.

```suggestion
        self.project().inner.poll_close(cx)
```

File: src/utils/io/writer_checksum.rs [hunk 3]
```diff
@@ -45,24 +46,24 @@ impl<D: Digest, W: AsyncWriteExt + Unpin> AsyncWrite for AsyncWriterWithChecksum
     }
 
     fn poll_flush(
-        mut self: std::pin::Pin<&mut Self>,
+        self: std::pin::Pin<&mut Self>,
         cx: &mut std::task::Context<'_>,
     ) -> std::task::Poll<std::io::Result<()>> {
-        Pin::new(&mut self.inner).poll_flush(cx)
+        let this = self.project();
+        this.inner.poll_flush(cx)
```
Comment (lemmih) [inline src/utils/io/writer_checksum.rs:53]:
```suggestion
        self.project().inner.poll_flush(cx)
```

**Review by lemmih — APPROVED** [review]

**Pull request merged**